src/main/pg/sdcard.c

```diff
@@ -57,6 +57,11 @@ void pgResetFn_sdcardConfig(sdcardConfig_t *config)
     config->mode = SDCARD_MODE_NONE;
 #endif
 
+#if defined(STM32H7) && defined(USE_SDCARD_SDIO) // H7 only for now, likely should be applied to F4/F7 too
```

Comment:
Defaulting this to SDIO will cause the firmware to hang for all targets that do not have an SDIO SD card and use the SDIO pins for something else.

Reply:
`USE_SDCARD_SDIO` is undefined by default, so I don't think this is an issue?

Reply:
In general, all features should be configurable at runtime.
This will allow Universal Targets to be used, making maintenance and release management scalable.
So the correct way to go on about this is for target maintainers of targets using SDIO for SD cards to set the appropriate mode in their target configuration - to the user this works in the same seamless way, and it allows other targets without SDIO SD card to use the same firmware binary.

Reply:
Is this something you're addressing in the H7 unified target? Do any other targets use SDIO at present?

Reply:
Yes, I am working on a Unified Target for STM32H743. In general, all features should be built to support Unified Targets.
WORMFC (F4) is the only non-H7 target to support SDIO at the moment. I did some testing with SDIO on F7 at some point, but back then it did not work, and I found some pointers on STM forums about the then current STM libraries for SDIO on F722 being buggy. F745 is working.

Reply:
I've just realised there was some old SPI DMA related legacy code for the SD card, so I've just pruned that out.

Reply:
Not sure how best to handle SPI vs SDIO detection for a unified target without running into the issue above. Thoughts?

Reply:
There is no need to make this auto-detected - any target that has either SPI or SDIO SD card slot will explicitly set the correct mode.
While auto-detection would be nice, I doubt that this is feasible, as the SDIO code in the STM library seems to hang when some of the SDIO pins are used for other purposes - it looks like this wasn't designed for this kind of auto-detection

Reply:
What about unified targets?

Reply:
They need to explicitly configure if they use SDIO or SPI for SD cards - just like it is done before this.

Reply:
OK, so line 60 should be replaced by a runtime check on `config->mode`. Any target wishing to use SDIO then needs to set `config->mode = SDCARD_MODE_SDIO;` in it's `targetConfiguration()` routine? I believe that's only called by `resetEEPROM()`. The setting you highlight above in `pgResetFn_sdcardConfig()` is called before the config is read from "EEPROM", so we need a another hook other than `targetConfiguration()` I suspect.

What target specific routine do we have in which `config->mode` could be set prior to reading the config?

Also note, in init.c:331 we have the following which I presume will cause the same problem and should again be conditional on `config->mode`? 

```
#if defined(STM32H7) && defined(USE_SDCARD_SDIO) // H7 only for now, likely should be applied to F4/F7 too
    sdioPinConfigure();
    SDIO_GPIO_Init();
#endif
```

Reply:
`detectHardwareRevision()` may be a suitable candidate.

Reply:
Unfortunately for a unified target we need to know how to configure the SDIO before the config that customised the unified config has been read.

Reply:
@mikeller Do you want this to handle the case where config is stored on the SD card? If so I think we’ll need to read flash based config first, then reload from SD.

Reply:
Yes, there's a catch-22 situation with targets that store config on SD card or external flash (the problem is the same on both).  You need to know the pins for the SD card or external flash chip before you can initialise the clocks and peripherals so the targets have to define them.  This doesn't fit well for the unified target approach, the H7 code SDIO and external flash support and the unified targets were developed at the same time so all the issues were not known at the time.